### PR TITLE
[dnn] fix high level api for python

### LINKED ISSUE #19346: Python binding for subclasses of TextDetectionModel crashes the program
Running the following lines with a proper model and frame crashes:

```
detector = cv.dnn_TextDetectionModel_DB(model=model)
result = detector.detectTextRectangles(frame)
```

Changing the export type of `TextDetectionModel` from
https://github.com/opencv/opencv/blob/49127b6a261ddb12a97c0d3febe0280fc0b59d7d/modules/dnn/include/opencv2/dnn/dnn.hpp#L1409
to
```
class CV_EXPORTS_W_SIMPLE TextDetectionModel : public Model
```

and recompiling and reinstalling the Python bindings fixes the crash for me
and `detector.detectTextRectangles(frame)` returns the correct results.

##### System information (version)
- OpenCV => 4.5.1
- Operating System / Platform => macOS 10.15.7
- Compiler => Apple clang version 12.0.0 (clang-1200.0.32.28)
- Python 3.9.1

 - [X] I report the issue, it's not a question
 - [X] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
 - [X] I updated to latest OpenCV version and the issue is still there

### LINKED ISSUE #19481: DeepLab v3 does not work with High Level API of Python
##### System information (version)

- OpenCV => 4.5.1
- Operating System / Platform => Windows 10 x64
- Compiler => Visual Studio 2019
- Python => 3.8

##### Detailed description

Python High Level API (<code>cv.dnn_SegmentationModel</code>) is not able to inference DeepLab v3 model.
However, It is able to inference with Python Low Level API (<code>cv.dnn.Net</code>) and C++ High Level API (<code>cv::dnn::SegmentationModel</code>) and C++ Low Level API (<code>cv::dnn::Net</code>).

* model download from open_model_zoo  
  https://github.com/openvinotoolkit/open_model_zoo/tree/2021.2/models/public/deeplabv3
  https://docs.openvinotoolkit.org/latest/omz_models_public_deeplabv3_deeplabv3.html

* convert optimize graph using this script
  https://github.com/opencv/opencv/issues/17362#issuecomment-634407860

##### Steps to reproduce

* Python High Level API (**Not Work**)
```py
import numpy as np
import cv2 as cv

# Read Classes Name
def read_class_name_list(file):
    classes = None
    with open(file, 'r') as f:
        classes = f.read().rstrip('\n').split('\n')
    return classes

# Get Classes Color
def get_class_colors(num):
    colors = []
    np.random.seed(0)
    for i in range(num):
        color = np.random.randint(0, 256, [3]).tolist()
        colors.append(color)
    return colors

if __name__ == '__main__':
    # Open Capture
    capture = cv.VideoCapture("bicycle.jpg")
    if not capture.isOpened():
        exit()

    # Create Model
    weights = "optimized_graph.pb"
    model = cv.dnn_SegmentationModel(weights)

    # Normalize Parameters
    scale = 0.007843
    size = (513, 513)
    mean = (127.5, 127.5, 127.5)
    swap = True
    crop = False
    model.setInputParams(scale, size, mean, swap, crop)

    # Read Classes Name and Colors
    names = "voc.names"
    classes = read_class_name_list(names)
    colors = get_class_colors(len(classes))

    while True:
        # Capture Frame
        result, image = capture.read()
        if result is False:
            cv.waitKey(0)
            break

        # Segmentation
        mask = model.segment(image) ## crash!!! ##

        # Generate Color Mask
        rows, cols = mask.shape
        color_mask = np.full(np.array([rows, cols, 3]), np.full(3, 0.), np.uint8)
        for y in range(rows):
            for x in range(cols):
                if mask[y, x] != 0: color_mask[y, x] = colors[mask[y, x]]
        color_mask = cv.resize(color_mask, (image.shape[1], image.shape[0]), cv.INTER_NEAREST)

        # Alpha Blend
        alpha = 0.5
        beta  = 1.0 - alpha
        cv.addWeighted(image, alpha, color_mask, beta, 0.0, image)

        # Show Image
        cv.imshow("segmentation", image)
        cv.imshow("mask", color_mask)
        key = cv.waitKey(10)
        if key == ord('q'):
            break
    
    cv.destroyAllWindows()
```

* C++ High Level API (Work Fine)
```cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <type_traits>
#include <opencv2/opencv.hpp>

// Read Class Name List
std::vector<std::string> readClassNameList( const std::string list_path )
{
    std::vector<std::string> classes;
    std::ifstream ifs( list_path );
    if( !ifs.is_open() ){
        return classes;
    }
    std::string class_name = "";
    while( std::getline( ifs, class_name ) ){
        classes.push_back( class_name );
    }
    return classes;
}

// Get Class Color Table for Visualize
std::vector<cv::Scalar> getClassColors( const int32_t number_of_colors )
{
    cv::RNG random;
    std::vector<cv::Scalar> colors;
    for( int32_t i = 0; i < number_of_colors; i++ ){
        cv::Scalar color( random.uniform( 0, 255 ), random.uniform( 0, 255 ), random.uniform( 0, 255 ) );
        colors.push_back( color );
    }
    return colors;
}

int main( int argc, char* argv[] )
{
    // Open Video Capture
    cv::VideoCapture capture = cv::VideoCapture( "bicycle.jpg" );
    if( !capture.isOpened() ){
        return -1;
    }

    // Create Model
    const std::string weights = "optimized_graph.pb";
    cv::dnn::SegmentationModel model = cv::dnn::SegmentationModel( weights );

    // Normalize Parameters
    constexpr double scale = 0.007843;
    const cv::Size size = cv::Size( 513, 513 );
    const cv::Scalar mean = cv::Scalar( 127.5, 127.5, 127.5 );
    constexpr bool swap = true;
    constexpr bool crop = false;
    model.setInputParams( scale, size, mean, swap, crop );

    // Read Classes Name and Colors
    const std::string list = "voc.names";
    const std::vector<std::string> classes = readClassNameList( list );
    const std::vector<cv::Scalar> colors = getClassColors( classes.size() );

    while( true ){
        // Capture Frame
        cv::Mat image;
        capture >> image;
        if( image.empty() ){
            cv::waitKey( 0 );
            break;
        }

        // Segmentation
        cv::Mat mask;
        model.segment( image, mask );

        // Generate Color Mask
        const int32_t rows = mask.rows;
        const int32_t cols = mask.cols;
        cv::Mat color_mask = cv::Mat::zeros( rows, cols, CV_8UC3 );
        for( int32_t y = 0; y < rows; y++ ){
            for( int32_t x = 0; x < cols; x++ ){
                const uint8_t class_id = mask.at<uint8_t>( y, x );
                if( class_id == 0 ) continue;
                const cv::Scalar color = colors[class_id];
                color_mask.at<cv::Vec3b>( y, x ) = cv::Vec3b( color[0], color[1], color[2] );
            }
        }
        cv::resize( color_mask, color_mask, image.size() );

        // Alpha Blend
        constexpr double alpha = 0.5;
        constexpr double beta = 1.0 - alpha;
        cv::addWeighted( image, alpha, color_mask, beta, 0.0, image );

        // Show Image
        cv::imshow( "segmentation", image );
        const int32_t key = cv::waitKey( 10 );
        if( key == 'q' ){
            break;
        }
    }

    cv::destroyAllWindows();

    return 0;
}

```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by UnaNancyOwen**
fix #19346
fix #19481

original opinion by @mikr. thanks! https://github.com/opencv/opencv/issues/19346#issue-788394632

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by VadimLevin** [comment]
@UnaNancyOwen Thank you for contribution. Can you add simple tests to verify Python bindings functionality for exported classes are fixed?

**Comment by UnaNancyOwen** [comment]
@VadimLevin Sorry, That could be difficult for me. I don't know how python / dnn-module testing in OpenCV.

**Comment by VadimLevin** [comment]
> @VadimLevin Sorry, That could be difficult for me. I don't know how python / dnn-module testing in OpenCV.

You can add a test to `modules/dnn/misc/python/test/test_dnn.py` as `test_high_level_api(self)` method to `class dnn_test`. To reproduce the issue before fix I think you can use models from `opencv_extra` repository. Before applied fix test should fail and after applied fix test should pass.

**Comment by UnaNancyOwen** [comment]
@VadimLevin I have written a test. Please check following.
And, I have one question. Do I need to upload model (.onnx) and reference (.npy) to opencv/opencv_extra?
```py
def test_textdetection_model_db(self):
    img_path = self.find_dnn_file("dnn/text_det_test1.jpg")
    weights = self.find_dnn_file("onnx/models/DB_TD500_resnet50.onnx", required=False)
    ref = np.load(self.find_dnn_file("dnn/DB_TD500_resnet50_prob.npy"))
    if weights is None:
        raise unittest.SkipTest("Missing DNN test files (onnx/models/DB_TD500_resnet50.onnx). Verify OPENCV_DNN_TEST_DATA_PATH configuration parameter.")

    frame = cv.imread(img_path)
    model = cv.dnn_TextDetectionModel_DB(weights)
    model.setInputScale(1.0 / 255.0)
    model.setInputSize(736, 736)
    model.setInputMean(122.67891434, 116.66876762, 104.00698793)

    out = model.predict(frame)
    normAssert(self, out, ref)
```

**Comment by VadimLevin** [comment]
> @VadimLevin I have written a test. Please check following.
> And, I have one question. Do I need to upload model (.onnx) and reference (.npy) to opencv/opencv_extra?

That's my suggestion to use model from `opencv_extra`, because the issue is related to the generated bindings and can be reproduced with any model. To avoid generating reference `npy` file you can use corresponding low level API  `cv::dnn::Net` and compare results - they should be close to each other.

**Comment by UnaNancyOwen** [comment]
> That's my suggestion to use model from opencv_extra, because the issue is related to the generated bindings and can be reproduced with any model. 

(I don't understand how it works, but) is it okay if the model is described in opencv_extra/testdata/dnn/download_models.py?
DB_TD500_resnet50.onnx is described in here.
https://github.com/opencv/opencv_extra/blob/4.5.1/testdata/dnn/download_models.py#L872-L876

> To avoid generating reference npy file you can use corresponding low level API cv::dnn::Net and compare results - they should be close to each other.

That's a good idea. I took that idea in test code.
```py
def test_textdetection_model(self):
    img_path = self.find_dnn_file("dnn/text_det_test1.png")
    weights = self.find_dnn_file("onnx/models/DB_TD500_resnet50.onnx", required=False)
    if weights is None:
        raise unittest.SkipTest("Missing DNN test files (onnx/models/DB_TD500_resnet50.onnx). Verify OPENCV_DNN_TEST_DATA_PATH configuration parameter.")

    frame = cv.imread(img_path)
    scale = 1.0 / 255.0
    size = (736, 736)
    mean = (122.67891434, 116.66876762, 104.00698793)

    model = cv.dnn_TextDetectionModel_DB(weights)
    model.setInputParams(scale, size, mean)
    out = model.predict(frame)

    net = cv.dnn.readNet(weights)
    blob = cv.dnn.blobFromImage(frame, scale, size, mean)
    net.setInput(blob)
    output_layers = net.getUnconnectedOutLayersNames()
    ref = net.forward(output_layers)

    normAssert(self, out, ref)
```

File: modules/dnn/misc/python/test/test_dnn.py [hunk 1]
```diff
@@ -197,6 +197,29 @@ def test_classification_model(self):
...
+            raise unittest.SkipTest("Missing DNN test files (onnx/models/DB_TD500_resnet50.onnx). Verify OPENCV_DNN_TEST_DATA_PATH configuration parameter.")
+
+        frame = cv.imread(img_path)
+        scale = 1.0 / 255.0
+        size = (736, 736)
+        mean = (122.67891434, 116.66876762, 104.00698793)
+
+        model = cv.dnn_TextDetectionModel_DB(weights)
+        model.setInputParams(scale, size, mean)
+        out = model.predict(frame)
```
Comment (alalek) [inline modules/dnn/misc/python/test/test_dnn.py:213]: This code shows that patch in .hpp file is not necessary to pass this test (test doesn't fail without the patch).

Please update test code (it should fail before and pass after the patch).

**Comment by alalek** [comment]
> To avoid generating reference npy file you can use corresponding low level API cv::dnn::Net and compare results

There is strong recommendation to avoid **accuracy** checking in Python (and others) bindings tests.
Check only parameters passing and types/shapes of returned results.

**Comment by UnaNancyOwen** [comment]
@alalek I will fix test to check type and shape of output.

**Review by alalek — APPROVED** [review]
Well done! Thank you :+1:

**Pull request merged**